masschallenge/impact-api
# Ac 4875: GET API Calls for Black Iron

Work done:
- Improves handling of OPTIONS calls by localizing the data in the helper objects.  More redundancy could be removed here, but it would be better to wait until we have decided on a standard per AC-4855.
- Provides read-only Program Family Resource (/api/v1/program_family)
- Provides read-only Confidential User Call (/api/v1/user/<id>/confidential) which is managed by a separate permission.
- Added Judging Round id and name information to User History call
- Added Newsletter events to User History call

To test:
- Program Family Resource:
-- Go to http://localhost:8000/api/v1/program_family/, check the OPTIONS, pick an "id" and go to http://localhost:8000/api/v1/program_family/<id>

- Confidential User Information:
-- Visit http://localhost:8000/api/v1/user/182/confidential/ and see that you get a permission denied warning.
-- Give your user the appropriate Django permission:
```
HQ-Dev-8:impact-api nathan$ make shell
Starting impactapi_mysql_1 ... done
Starting impactapi_redis_1 ... done
Python 3.6.2 (default, Jul 24 2017, 19:47:39) 
Type "copyright", "credits" or "license" for more information.

IPython 4.0.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import impact.models as im

In [2]: user = im.User.objects.get(id=13514)

In [3]: from django.contrib.auth.models import Group

In [4]: from django.conf import settings

In [5]: group = Group.objects.create(name=settings.V1_CONFIDENTIAL_API_GROUP)

In [6]: user.groups.add(group)

In [7]: user.save()
```
-- See this http://localhost:8000/api/v1/user/182/confidential/ succeed
-- Note that there is currently only information related to experts in this call.

- Judging Round information in User History call
-- Find a user who has been a confirmed judge, say the user with id=182
-- Make the appropriate User History call, e.g., http://localhost:8000/api/v1/user/182/history/
-- Search for the string "judging_round_name"

- Newsletter Events in User History call
-- On the above page search for "newsletter_name"

- Make sure the OPTIONS calls are reasonable for the above GET calls.
